smart_contract/executor/src/lib.rs

```diff
@@ -172,34 +182,106 @@ macro_rules! deny {
     }};
 }
 
-/// Collection of all permission tokens defined by the executor
-#[derive(Debug, Clone, Default)]
-pub struct PermissionSchema(Vec<PermissionId>, MetaMap);
+/// Convert a type to and from [`ExecutorDataModelObject`].
+pub trait ConvertDataModelObject: Serialize + DeserializeOwned {
+    /// Target object.
+    type Object: ExecutorDataModelObject;
+
+    /// Implementor's type ID as specified in the data model schema.
+    fn id() -> <Self::Object as ExecutorDataModelObject>::Id;
 
-impl PermissionSchema {
-    /// Remove permission token from this collection
-    pub fn remove<T: permission::Token>(&mut self) {
-        let to_remove = <T as permission::Token>::name();
+    /// Try to convert an object to the type
+    /// # Errors
+    /// - If object id doesn't match with type id
+    /// - If fails to deserialize object payload into the type
+    fn try_from_object(object: &Self::Object) -> Result<Self, ConvertDataModelObjectError> {
+        let expected_id = Self::id();
 
-        if let Some(pos) = self.0.iter().position(|token_id| *token_id == to_remove) {
-            self.0.remove(pos);
-            <T as iroha_schema::IntoSchema>::remove_from_schema(&mut self.1);
+        if *object.id() != expected_id {
+            return Err(ConvertDataModelObjectError::Id(object.id().clone().into()));
         }
+
+        object
+            .payload()
+            .deserialize()
+            .map_err(ConvertDataModelObjectError::Deserialize)
+    }
+
+    /// Convert the type into its object representation
+    fn into_object(self) -> Self::Object {
+        let id = Self::id();
+
+        let payload = DebugExpectExt::dbg_expect(
+            serde_json::to_value(self),
+            "failed to serialize concrete data model entity; this is a bug",
+        );
+
+        Self::Object::new(id, JsonString::from(&payload))
+    }
+}
+
+/// An error that might occur while converting an [`ExecutorDataModelObject`]
+/// into a native executor type.
+#[derive(Debug)]
+pub enum ConvertDataModelObjectError {
+    /// Unexpected object id
+    Id(data_model::prelude::Name),
+    /// Failed to deserialize object payload
+    Deserialize(serde_json::Error),
+}
+
+/// A convenience to build [`ExecutorDataModel`] from within the executor
+#[derive(Debug, Clone, Default)]
```

Comment:
```suggestion
#[derive(Debug, Clone)]
```

I'd keep it without `Default` so as not to confuse with `with_default_permissions`